In [1]:
import requests
import pandas as pd
import time

EDA - Alojamientos
Pasos a seguir.
* El alojamiento lo he obtenido usando la API de Booking. A través de esta pude extraer los datos con los que estaré trabajando a continuación. En cuanto a los alojamientos, estos ya se han unido en un solo Data Frame. De esta manera, podré comparar en el mismo las cosas directamente. 
He hecho cierta limpieza a medida que iba dando forma al Data Frame, pero aquí se evaluará mejor formato y descripción de los datos.

In [45]:
df_florencia_lisboa = pd.read_csv('../datos/alojamientos_combinados.csv')

In [46]:
df_florencia_lisboa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Ciudad               40 non-null     object 
 1   Hotel                40 non-null     object 
 2   Cantidad de reviews  40 non-null     int64  
 3   Calificación         40 non-null     float64
 4   Precio               40 non-null     float64
 5   Entrada desde        40 non-null     object 
 6   Entrada hasta        40 non-null     object 
 7   Salida desde         40 non-null     object 
 8   Salida hasta         40 non-null     object 
dtypes: float64(2), int64(1), object(6)
memory usage: 2.9+ KB


Tengo que convertir las horas en formato datetime. Puedo ver que no tengo nulos

In [62]:
df_florencia_lisboa.duplicated().sum()

np.int64(0)

In [47]:
df_florencia_lisboa.head()

Ciudad                                          Hotel  \
0  Florence  Florence Heart Uffizi - Renovated and central   
1  Florence                         Appartamento la Ninfea   
2  Florence                        BorgoDeGreci Apartments   
3  Florence                        B&B Le Stanze del Duomo   
4  Florence        Adler Cavalieri Hotel-Private Spa & Gym   

   Cantidad de reviews  Calificación       Precio Entrada desde Entrada hasta  \
0                   70           9.1   585.000000         15:00         20:00   
1                  195           8.9  1842.380000         14:00         19:00   
2                  468           8.7   590.630000         14:00         00:00   
3                 6580           8.5   619.304625         14:00         21:00   
4                 3628           8.5   722.000000         15:00         00:00   

  Salida desde Salida hasta  
0        08:00        10:00  
1        07:00        10:00  
2        06:00        10:00  
3        08:00        11:00  
4        00:00        12:00

In [48]:
df_florencia_lisboa['Precio'] = df_florencia_lisboa['Precio'].round(2)
# quitamos el exceso de ceros en el precio
#  redondeo los precios a dos decimales

In [49]:
df_florencia_lisboa.head(1)

Ciudad                                          Hotel  \
0  Florence  Florence Heart Uffizi - Renovated and central   

   Cantidad de reviews  Calificación  Precio Entrada desde Entrada hasta  \
0                   70           9.1   585.0         15:00         20:00   

  Salida desde Salida hasta  
0        08:00        10:00

In [50]:
df_florencia_lisboa.head()

Ciudad                                          Hotel  \
0  Florence  Florence Heart Uffizi - Renovated and central   
1  Florence                         Appartamento la Ninfea   
2  Florence                        BorgoDeGreci Apartments   
3  Florence                        B&B Le Stanze del Duomo   
4  Florence        Adler Cavalieri Hotel-Private Spa & Gym   

   Cantidad de reviews  Calificación   Precio Entrada desde Entrada hasta  \
0                   70           9.1   585.00         15:00         20:00   
1                  195           8.9  1842.38         14:00         19:00   
2                  468           8.7   590.63         14:00         00:00   
3                 6580           8.5   619.30         14:00         21:00   
4                 3628           8.5   722.00         15:00         00:00   

  Salida desde Salida hasta  
0        08:00        10:00  
1        07:00        10:00  
2        06:00        10:00  
3        08:00        11:00  
4        00:00        12:00

In [ ]:
df_florencia_lisboa.info()

In [52]:
df_florencia_lisboa.shape

(40, 9)

In [53]:
df_florencia_lisboa.columns

Index(['Ciudad', 'Hotel', 'Cantidad de reviews', 'Calificación', 'Precio',
       'Entrada desde', 'Entrada hasta', 'Salida desde', 'Salida hasta'],
      dtype='object')

In [54]:
df_florencia_lisboa.describe()

Cantidad de reviews  Calificación       Precio
count            40.000000     40.000000    40.000000
mean           2347.700000      8.652500   641.281500
std            1961.433244      0.479309   341.147894
min               5.000000      7.200000    81.870000
25%             821.750000      8.400000   484.857500
50%            1557.000000      8.700000   587.815000
75%            3835.750000      8.925000   704.172500
max            6580.000000      9.500000  1842.380000

In [55]:
df_florencia_lisboa.describe(include='O')

Ciudad                                          Hotel Entrada desde  \
count         40                                             40            40   
unique         2                                             40             7   
top     Florence  Florence Heart Uffizi - Renovated and central         14:00   
freq          20                                              1            19   

       Entrada hasta Salida desde Salida hasta  
count             40           40           40  
unique             7            7            5  
top            00:00        00:00        12:00  
freq              27           24           16

In [58]:
df_florencia_lisboa.groupby('Ciudad')['Precio'].mean()

Ciudad
Florence    727.991
Lisbon      554.572
Name: Precio, dtype: float64

El precio medio de hospedaje para estas fechas en Lisboa es de 554€ mientras que en Florencia es de 727€.

In [60]:
df_florencia_lisboa.groupby('Ciudad').describe().T

Ciudad                        Florence       Lisbon
Cantidad de reviews count    20.000000    20.000000
                    mean   1650.750000  3044.650000
                    std    1678.261942  2014.386710
                    min      70.000000     5.000000
                    25%     448.750000  1107.250000
                    50%    1253.500000  3312.000000
                    75%    2116.750000  4820.500000
                    max    6580.000000  5874.000000
Calificación        count    20.000000    20.000000
                    mean      8.775000     8.530000
                    std       0.378188     0.544929
                    min       7.800000     7.200000
                    25%       8.500000     8.200000
                    50%       8.800000     8.600000
                    75%       9.025000     8.825000
                    max       9.400000     9.500000
Precio              count    20.000000    20.000000
                    mean    727.991000   554.572000
                    std     412.611753   229.808998
                    min     145.000000    81.870000
                    25%     555.922500   454.050000
                    50%     604.965000   514.710000
                    75%     704.172500   672.975000
                    max    1842.380000  1001.710000

En cuanto a la calificación promedio, en base a 10, encontré que Florencia tiene una media de 8.77 mientras que Lisboa tiene una de 8.53.
En cuanto a la cantidad de reviews, Florencia tiene una media de 1650 que se queda muy distante de la media de Lisboa que se presenta en 3044.
* Considerando que los precios son más asequibles para Lisboa, puede que la carga histórica y cultural de Florencia eleve más los precios y que por ende más personas decidan asistir en viajes low-cost a Lisboa.
* A esto se suma que el mínimo de hospedaje en Lisboa empieza en 81€ mientras que en Florencia es 145€. También podemos observa que la mediana en ambos lugares no es tan distante, Florencia: 604.96€ y Lisboa: 514.71€. Sin embargo, tenemos un precio máximo de 1842€ en Florencia y 1001€ en Lisboa. Recalco que la riqueza cultural e histórica de Florencia juega un papel importante ya que algunos alojamientos están bastante cerca de lugares memorables de la ciudad.

---
* Vuelos a Florencia*

A continuación, busco la información sobre los vuelos de MAD-FLR y FLR-MAD durante las fechas del 25-10-2024 al 28-10-2024.

In [63]:
df_vuelos_florencia = pd.read_csv('../datos/vuelos_florencia_itinerario.csv')

In [67]:
df_vuelos_florencia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Costo            20 non-null     float64
 1   ID Salida        20 non-null     object 
 2   Salida           20 non-null     object 
 3   ID Llegada       20 non-null     object 
 4   Llegada          20 non-null     object 
 5   Tiempo           20 non-null     int64  
 6   Paradas          20 non-null     int64  
 7   Aerolínea        20 non-null     object 
 8   Hora de Salida   20 non-null     object 
 9   Hora de Llegada  20 non-null     object 
dtypes: float64(1), int64(2), object(7)
memory usage: 1.7+ KB


In [104]:
df_vuelos_florencia['Hora de Salida'].duplicated().sum()

np.int64(10)

In [105]:
df_vuelos_florencia['Hora de Llegada'].duplicated().sum()

np.int64(11)

En este caso, para Florencia tengo varios horarios duplicados pero destaco el tema de jugar con las combinaciones. Aunque el vuelo de ida pueda repetirse, este tiene una combinación con el vuelo de regreso distinto en todos los caso. Pasa lo mismo con los vuelos duplicados de regreso. 

In [101]:
df_vuelos_florencia

Costo ID Salida    Salida ID Llegada   Llegada  Tiempo  Paradas  \
0   428.00       MAD    Madrid        FLR  Florence     140        0   
1   428.00       FLR  Florence        MAD    Madrid     150        0   
2   429.00       MAD    Madrid        FLR  Florence     140        0   
3   429.00       FLR  Florence        MAD    Madrid     150        0   
4   290.00       MAD    Madrid        FLR  Florence     250        1   
5   290.00       FLR  Florence        MAD    Madrid     150        0   
6   281.68       MAD    Madrid        FLR  Florence     300        1   
7   281.68       FLR  Florence        MAD    Madrid     284        1   
8   451.98       MAD    Madrid        FLR  Florence     140        0   
9   451.98       FLR  Florence        MAD    Madrid     150        0   
10  438.00       MAD    Madrid        FLR  Florence     140        0   
11  438.00       FLR  Florence        MAD    Madrid     150        0   
12  295.00       MAD    Madrid        FLR  Florence     380        1   
13  295.00       FLR  Florence        MAD    Madrid     150        0   
14  308.00       MAD    Madrid        FLR  Florence     450        1   
15  308.00       FLR  Florence        MAD    Madrid     150        0   
16  316.00       MAD    Madrid        FLR  Florence     300        1   
17  316.00       FLR  Florence        MAD    Madrid     150        0   
18  340.00       MAD    Madrid        FLR  Florence     415        1   
19  340.00       FLR  Florence        MAD    Madrid     150        0   

           Aerolínea       Hora de Salida      Hora de Llegada  
0   Vueling Airlines  2024-10-25T09:55:00  2024-10-25T12:15:00  
1   Vueling Airlines  2024-10-28T12:35:00  2024-10-28T15:05:00  
2   Vueling Airlines  2024-10-25T15:35:00  2024-10-25T17:55:00  
3   Vueling Airlines  2024-10-28T12:35:00  2024-10-28T15:05:00  
4   Vueling Airlines  2024-10-25T17:20:00  2024-10-25T21:30:00  
5   Vueling Airlines  2024-10-28T12:35:00  2024-10-28T15:05:00  
6        ITA Airways  2024-10-25T05:50:00  2024-10-25T10:50:00  
7        ITA Airways  2024-10-28T19:15:00  2024-10-28T23:59:00  
8   Vueling Airlines  2024-10-25T09:55:00  2024-10-25T12:15:00  
9   Vueling Airlines  2024-10-28T06:15:00  2024-10-28T08:45:00  
10  Vueling Airlines  2024-10-25T15:35:00  2024-10-25T17:55:00  
11  Vueling Airlines  2024-10-28T06:15:00  2024-10-28T08:45:00  
12        Air Europa  2024-10-25T15:10:00  2024-10-25T21:30:00  
13  Vueling Airlines  2024-10-28T12:35:00  2024-10-28T15:05:00  
14       ITA Airways  2024-10-25T07:05:00  2024-10-25T14:35:00  
15  Vueling Airlines  2024-10-28T12:35:00  2024-10-28T15:05:00  
16       ITA Airways  2024-10-25T05:50:00  2024-10-25T10:50:00  
17  Vueling Airlines  2024-10-28T12:35:00  2024-10-28T15:05:00  
18       ITA Airways  2024-10-25T11:35:00  2024-10-25T18:30:00  
19  Vueling Airlines  2024-10-28T12:35:00  2024-10-28T15:05:00

In [68]:
df_vuelos_florencia.shape

(20, 10)

In [70]:
df_vuelos_florencia.columns

Index(['Costo', 'ID Salida', 'Salida', 'ID Llegada', 'Llegada', 'Tiempo',
       'Paradas', 'Aerolínea', 'Hora de Salida', 'Hora de Llegada'],
      dtype='object')

In [78]:
df_vuelos_florencia['Aerolínea'].value_counts(normalize=True) * 100

Aerolínea
Vueling Airlines    70.0
ITA Airways         25.0
Air Europa           5.0
Name: proportion, dtype: float64

Puedo observar que de las tres aerolíneas, Vueling tiene el 70% de los vuelos que se ofrecen en estas fechas de ida y vuelta. Seguidos de ITA Airways.

In [92]:
df_vuelos_florencia.duplicated().sum()

np.int64(0)

In [69]:
df_vuelos_florencia.describe()

Costo      Tiempo   Paradas
count   20.00000   20.000000  20.00000
mean   357.76600  214.450000   0.35000
std     68.19356  103.467501   0.48936
min    281.68000  140.000000   0.00000
25%    295.00000  150.000000   0.00000
50%    328.00000  150.000000   0.00000
75%    429.00000  288.000000   1.00000
max    451.98000  450.000000   1.00000

En cuanto a media del costo de vuelos, ida y vuelta, encontré un promedio de 357€ para Florencia.
Una duración promedio de vuelo de 214 minutos, aunque la mediana concentra los valores en 150 minutos. Lo que me indica que aquellos vuelos que tienen escala, extienden la duración del vuelo. Esto se puede observa al señalar que el tiempo máximo es de 450 minutos. Sin embargo, lo bueno es que aquellos vuelos que tienen parada se limitan a solo 1.

A partir de aquí, empiezo a ver los datos para los vuelos a Lisboa.

In [64]:
df_vuelos_lisboa = pd.read_csv('../datos/vuelos_lisboa_itinerario.csv')

In [97]:
df_vuelos_lisboa

Costo ID Salida  Salida ID Llegada Llegada  Tiempo  Paradas Aerolínea  \
0   154.36       MAD  Madrid        LIS  Lisbon      80        0   easyJet   
1   154.36       LIS  Lisbon        MAD  Madrid      80        0   easyJet   
2   160.41       MAD  Madrid        LIS  Lisbon      85        0    Iberia   
3   160.41       LIS  Lisbon        MAD  Madrid      85        0    Iberia   
4   130.80       MAD  Madrid        LIS  Lisbon      85        0   Ryanair   
5   130.80       LIS  Lisbon        MAD  Madrid      80        0   Ryanair   
6   146.36       MAD  Madrid        LIS  Lisbon      80        0   easyJet   
7   146.36       LIS  Lisbon        MAD  Madrid      85        0   easyJet   
8   146.36       MAD  Madrid        LIS  Lisbon      80        0   easyJet   
9   146.36       LIS  Lisbon        MAD  Madrid      85        0   easyJet   
10  187.30       MAD  Madrid        LIS  Lisbon      85        0   easyJet   
11  187.30       LIS  Lisbon        MAD  Madrid      80        0   Ryanair   
12  173.41       MAD  Madrid        LIS  Lisbon      85        0    Iberia   
13  173.41       LIS  Lisbon        MAD  Madrid      85        0    Iberia   
14  151.98       MAD  Madrid        LIS  Lisbon      85        0   Ryanair   
15  151.98       LIS  Lisbon        MAD  Madrid      80        0   easyJet   
16  143.98       MAD  Madrid        LIS  Lisbon      85        0   Ryanair   
17  143.98       LIS  Lisbon        MAD  Madrid      85        0   easyJet   
18  129.30       MAD  Madrid        LIS  Lisbon      80        0   easyJet   
19  129.30       LIS  Lisbon        MAD  Madrid      80        0   Ryanair   

         Hora de Salida      Hora de Llegada  
0   2024-10-25T22:35:00  2024-10-25T22:55:00  
1   2024-10-28T15:05:00  2024-10-28T17:25:00  
2   2024-10-25T23:00:00  2024-10-25T23:25:00  
3   2024-10-28T12:30:00  2024-10-28T14:55:00  
4   2024-10-25T21:35:00  2024-10-25T22:00:00  
5   2024-10-28T22:00:00  2024-10-29T00:20:00  
6   2024-10-25T22:35:00  2024-10-25T22:55:00  
7   2024-10-28T20:00:00  2024-10-28T22:25:00  
8   2024-10-25T22:35:00  2024-10-25T22:55:00  
9   2024-10-28T07:15:00  2024-10-28T09:40:00  
10  2024-10-25T09:55:00  2024-10-25T10:20:00  
11  2024-10-28T22:00:00  2024-10-29T00:20:00  
12  2024-10-25T23:00:00  2024-10-25T23:25:00  
13  2024-10-28T18:45:00  2024-10-28T21:10:00  
14  2024-10-25T21:35:00  2024-10-25T22:00:00  
15  2024-10-28T15:05:00  2024-10-28T17:25:00  
16  2024-10-25T21:35:00  2024-10-25T22:00:00  
17  2024-10-28T20:00:00  2024-10-28T22:25:00  
18  2024-10-25T22:35:00  2024-10-25T22:55:00  
19  2024-10-28T22:00:00  2024-10-29T00:20:00

In [72]:
df_vuelos_lisboa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Costo            20 non-null     float64
 1   ID Salida        20 non-null     object 
 2   Salida           20 non-null     object 
 3   ID Llegada       20 non-null     object 
 4   Llegada          20 non-null     object 
 5   Tiempo           20 non-null     int64  
 6   Paradas          20 non-null     int64  
 7   Aerolínea        20 non-null     object 
 8   Hora de Salida   20 non-null     object 
 9   Hora de Llegada  20 non-null     object 
dtypes: float64(1), int64(2), object(7)
memory usage: 1.7+ KB


In [94]:
df_vuelos_lisboa['Aerolínea'].value_counts(normalize=True) * 100

Aerolínea
easyJet    50.0
Ryanair    30.0
Iberia     20.0
Name: proportion, dtype: float64

En relación a los vuelos, cuento con 3 opciones de aerolíneas: easyJet (50%), Ryanair (30%) e Iberia (20%). Esto significa que easyJet tiene la mitad de los vuelos de esta ruta MAD-FLR/FLR-MAD.

In [81]:
df_vuelos_lisboa['Costo'] = df_vuelos_lisboa['Costo'].round(2)

In [89]:
df_vuelos_lisboa.describe().round(2)

Costo  Tiempo  Paradas
count   20.00   20.00     20.0
mean   152.43   82.75      0.0
std     17.39    2.55      0.0
min    129.30   80.00      0.0
25%    143.98   80.00      0.0
50%    149.17   85.00      0.0
75%    160.41   85.00      0.0
max    187.30   85.00      0.0

El costo promedio de un vuelo a Lisboa se encuentra en 152.42, cabe resaltar que la mediana de esta ruta es incluso menor (149.17€). Esto muestra que esta es una ruta más accesible y cuenta con menos disturbios, ya que no cuenta con escalas en ninguna de las rutas propuestas.
El tiempo es bastante uniforme por la razón de que no tiene escalas, el timpo promedio es de 82.75 que se traduce a 1h 23 min.

In [96]:
df_vuelos_lisboa.describe(include='O')

ID Salida  Salida ID Llegada Llegada Aerolínea       Hora de Salida  \
count         20      20         20      20        20                   20   
unique         2       2          2       2         3                   10   
top          MAD  Madrid        LIS  Lisbon   easyJet  2024-10-25T22:35:00   
freq          10      10         10      10        10                    4   

            Hora de Llegada  
count                    20  
unique                   10  
top     2024-10-25T22:55:00  
freq                      4

In [99]:
df_vuelos_lisboa[df_vuelos_lisboa.duplicated()]

Costo ID Salida  Salida ID Llegada Llegada  Tiempo  Paradas Aerolínea  \
8  146.36       MAD  Madrid        LIS  Lisbon      80        0   easyJet   

        Hora de Salida      Hora de Llegada  
8  2024-10-25T22:35:00  2024-10-25T22:55:00

Se puede observar que existe un duplicado, que se repite 4 veces. Este es el caso que muestro arriba: vuelo MAD-LIS, con easyJet, en horario de 22:35 a 22:55. Esto sucede porque este mismo vuelo de ida, cuenta con 4 opciones de vuelo de regreso LIS-MAD. Por lo que, a pesar de estar duplicado, tiene sentido que se mantenga. Además, al ser una combinación de vuelos, la vuelta tiene horarios distintos por lo que se refleja también en su precio. 